[Johannes] iP

jwdavis0200 opened the pull request:
# Duker

> “To-Do lists help us break life into small steps.” — Randy Pausch

Duker isn't like any other to-do list. You don't have to install anything on your computer!
Simply,
1. Download it [here](https://github.com/jwdavis0200/ip_own/releases/tag/Duke_releaseV0.3).
2. Double-Click
3. Start slaying!

## What features do I get with Duker?
- Remove your to-dos. *Be Gone!!*
- Add them! *Remember adding is just as important as removing!*.
- Find your tasks. *We know it gets crazy. :"^)*
- List them out, mark them, unmark them and *_MORE_*!

Future features to be Added:
[ ] Make it look better for your eyes.
[ ] Make it ~faster~ **EVEN** *faster*
[ ] Many mOre, stay tuned! 👍 

If you are a Java Programmer, take a look at the source code to see how OOP this is!!
This is our main.

```
public static void main(String[] args) {
        try {
            new Duke("storage.txt").run();
        } catch (DukeException de) {
            System.out.println("--------------------------------------\n");
            System.out.println(de.getMessage());
            System.out.println("--------------------------------------\n");
            main(args);
        } catch (IOException io) {
            System.out.println("--------------------------------------\n");
            System.out.println(io.getMessage());
            System.out.println("--------------------------------------\n");
        }
    }
```

Commit:
Add Gradle support

jwdavis0200 committed:
Added Level-1 functionality, Greet, Echo, Exit

jwdavis0200 committed:
Fixed cosmetic bugs and a technical bug where whitespace will result in a new prompt feedback

jwdavis0200 committed:
Added bye method for more general implementation for the end of operation of chatbot to minimise code repetition.

jwdavis0200 committed:
Added storeman method for storing of lists of input and outputting them as a list when the string list is passed as an input

jwdavis0200 committed:
Added Task Class to encapsulate tasks passed as input.
Added checklist functionality and modified storeman method to accomodate new function.
Added listOut method to generalise listing out of tasks, to minimise code repetition.

jwdavis0200 committed:
Added three subclasses of Task; Deadline, Event, Todo

jwdavis0200 committed:
Removed reduncant, unused checklist method

jwdavis0200 committed:
Edited Todo Class to meet specifications of project

jwdavis0200 committed:
Added support for tracking of three types of classes, Todo, Event and Deadlines

jwdavis0200 committed:
Replaced contains method with startsWith for program correctness and flexibility

jwdavis0200 committed:
Added three subclasess of Task; Todo, Event, Deadline

jwdavis0200 committed:
removed redundant, not used checklist method(checklist functionality implemented within storeman method). replaced contains with startsWith where appropariate to ensure correctness of program when running.

jwdavis0200 committed:
Various bug fix for Level-3

jwdavis0200 committed:
Change to Guard clause to ensure correctness and better flexibility of program

jwdavis0200 committed:
Bug Fix for Event TAG attribute

jwdavis0200 committed:
Added support for subclasses of Task; for Todo, Event, Deadline tracking

jwdavis0200 committed:
Added DukeException for throwing Duke-caused Exceptions

jwdavis0200 committed:
Added feature for storeman method to throw DukeException if input is unknown or in the wrong format, to be caught by main. Main will restart duke upon catching the exception and print the message of the error to console right before.

jwdavis0200 committed:
Added to gitignore

jwdavis0200 committed:
added to gitignore

jwdavis0200 committed:
Support for delete added

jwdavis0200 committed:
updated gitignore, added test-ui-test(but to be configured to be ignored)

jwdavis0200 committed:
removed debug statement

jwdavis0200 committed:
update files to be ignored

jwdavis0200 committed:
Updated input, expected and actual as well as runtest.bat to automate the process of testing

jwdavis0200 committed:
Merge branch 'master' into temp

jwdavis0200 committed:
Merge pull request #1 from jwdavis0200/temp

Merge pull request 1 from temp branch to master branch

jwdavis0200 committed:
Add save and load functionality to Duke and minor changes to String representation of  Task and Deadline

jwdavis0200 committed:
Add storage.txt as a file to be ignored by .gitignore

storage.txt is a cached file by Duke to save the list of Tasks by user in previous run.

storage.txt changed every run and there's no need to track the file in repo.

Hence, should be include in .gitignore as a file to be ignored by git tracking

jwdavis0200 committed:
Merge pull request #2 from nus-cs2103-AY2223S1/add-gradle-support

Add Gradle support

jwdavis0200 committed:
Enable Duke to process inputs by users for time-related inputs to be processed as LocalDateTime for Deadline and Event task that has time-related attributes

Duke can only process time-related attributes (due for deadline and duration for event) as String

This limits the manipulation of these field and makes it difficult for printing the inputs in a different format.

Hence, we should add this functionality to improve the versality of Duke such as to enable it to print out inputs in a different format.

jwdavis0200 committed:
Merge branch 'branch-level-7'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java
#	src/main/java/Task.java

jwdavis0200 committed:
Modify toString methods of Deadline and Event

toString methods lacks spacing

To make it nicer, we should add the whitespace for easier readability of output

Hence this commit will improve readability for client

jwdavis0200 committed:
Merge branch 'master' of https://github.com/jwdavis0200/ip_own

jwdavis0200 committed:
Remove wildcard import in favour of specific imports of classes

IDE reccomended wildcard import for Java.io since there were many classes from this package used in the program.

We should always avoid using wildcard imports in accordance to the Java Code Standard in CS2103T.

We have to replace the wildcard import with specific import statements for each specific packages.

jwdavis0200 committed:
Add more OOP principles

Currently Duke program handles all UI and storage and TaskList functionality mostly by itself.

We should add more OOP principle that makes sense to support the Java OOP paradigm as well as to improve code neatness and reusability via encapsulation and OOP Principles.

Hence the addition of Storage, UI and TaskList classes.

jwdavis0200 committed:
Add Parser class

Duke class was responsible for processing client inputs and categorising the tasks into the different types for creation.

It would be better to create a class specific for this purpose, letting the Duke class mostly only call on other classes and to do minimal conditionals.

The parser class is addded to process user inputs and to categorise the commands for different Duke functionality.

jwdavis0200 committed:
Modify Duke, Storage, TaskList, Ui class to support OOP principles with the addition of Parser class.

Parser class requires the restructuring of some code in these utility classes as well as Duke. Some bug fixing is also done.

The modifications allow Duke class to call on these utility classes, including the Parser class to follow better OOP paradigm.

jwdavis0200 committed:
Remove some import messages.

Some import messages in Duke class is no longer needed.

Needless to keep these import messages, maybe inefficient and decreases readability.

Should remove these messages since they are no longer needed.

jwdavis0200 committed:
Add folders and package the classes.

The classes are not packaged and organised.

We need to organise our files better, into duke package, utility package and dukeexception package.

Hence we add package statements for each file according to where they are organised and import package statements accordingly.

jwdavis0200 committed:
Add basic JUnit tests for Event and Todo classes

No JUnit test to test correctness of classes.

We want to test the toString methods of Event and Todo classes to ensure correctness.

Added JUnit tests for the toString methods of Event and Todo classes using JUnit.

jwdavis0200 committed:
Add find functionality to Duke

Currently, Duke does not have functionality to find matching tasks in the TaskList when requested by client

We want to allow client to find their matching tasks when requested.

Added conditionals in Parser and methods in TaskList and Ui classes to support this functionality.

jwdavis0200 committed:
Add task to build.gradle file to automate building of fat jar and remove method not used from Ui class.

askForPath is a method from the Ui class which is not used. There is no task to create a fat jar via build.gradle.

askForPath affects readability and adds extra lines of unnecessary codes. We also need a task to create a fat jar via build.gradle.

We should remove askForPath method sicne it is nto used and add task to build.gradle in order to automate the creation of fat jar file.

jwdavis0200 committed:
Remove method add from Storage class, implement add method in TaskList instead. Added Java documentation.

add method is in Storage class but it deals with adding tasks into the task list. There is no Java documentation.

TaskList class should be handling this operaton. Java documentations should also be added for the classes.

Transfer the add method to the TaskList class from the Storage class. Added Java documentation for the classes of the program.

jwdavis0200 committed:
Fix code to meet Checkstyle.

Previously, Checkstyle was not implemented to check for coding standard violations.

There are many Checkstyle coding violations upon implementing the Checkstyle Feature via Gradle.

Fixed many of such violations to meet the Checkstyle coding standards.

sugiyem commented on src/main/java/dukechatbot/duke/Duke.java:
```diff
@@ -0,0 +1,87 @@
+    private TaskList tasks;
+    private Ui ui;
+    private ArrayList<Task> al;
+    private Parser parse;
```
Maybe the variable al and parse can be renamed to a more descriptive object name?

sugiyem commented on src/main/java/dukechatbot/utility/Deadline.java:
```diff
@@ -0,0 +1,41 @@
+    /**
+     * Defines the Date time format to read in the commands passed by client for processing.
+     */
+    protected static final DateTimeFormatter DTF = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm");
```
Nice abstraction!

sugiyem commented on src/main/java/dukechatbot/duke/Duke.java:
```diff
@@ -0,0 +1,87 @@
+        try {
+            new Duke("storage.txt").run();
+        } catch (DukeException de) {
+            System.out.println("--------------------------------------\n");
```
This dashed line is used multiple times. Maybe it can be defined as a constant?

sugiyem commented on src/main/java/dukechatbot/utility/Storage.java:
```diff
@@ -0,0 +1,83 @@
+package dukechatbot.utility;
+import java.io.*;
```
Maybe you can list all the imported classes explicitly?

paullee18 replied:
Could import classes explicitly

sugiyem commented on src/main/java/dukechatbot/utility/TaskList.java:
```diff
@@ -0,0 +1,129 @@
+     * Encapsulates the array list associated with the instance of
+     * TaskList
+     */
+    private ArrayList<Task> tl;
```
Maybe can rename this variable to make it more descriptive?

sugiyem commented on src/main/java/dukechatbot/utility/Parser.java:
```diff
@@ -0,0 +1,96 @@
+                        if (uncap.startsWith("deadline")) {
+                            int idOfSlash = str.indexOf('/');
+                            if (idOfSlash - 9 == 0) {
+                                this.ui.showDescEmptyError("deadline");
```
I think this deeply nested if-else statement make the code quiet hard to read

sugiyem commented on src/main/java/dukechatbot/utility/Ui.java:
```diff
@@ -0,0 +1,172 @@
+    /**
+     * Defines the array list that is associated with the program run.
+     */
+    private ArrayList<Task> tl;
```
I think this variable can also be renamed to something more descriptive.

sugiyem reviewed (commented):
In general, I think your code looks good already. There's some coding standard violations listed below, but other than that, LGTM!

jwdavis0200 committed:
Add incomplete Basic GUI functionality to Duke

Duke was a CLI application.

To make it more appealing, we should implement a basic GUI.

Added basic GUI functionality using JavaFx.

jwdavis0200 committed:
Document classes implementing UI.

There was no documentation for the new classes previously.

We need documentation to explain the new classes to fellow programmers.

Documented the different classes implementing the GUI.

paullee18 commented on src/main/java/dukechatbot/duke/Duke.java:
```diff
@@ -0,0 +1,98 @@
+    private Storage storage;
+    private TaskList tasks;
+    private Ui ui;
+    private ArrayList<Task> al;
```
Perhaps there's no need for another ArrayList when theres already a TaskList

paullee18 commented on src/main/java/dukechatbot/utility/Parser.java:
```diff
@@ -0,0 +1,101 @@
+                    if (uncap.startsWith("unmark")) {
+                        int i = Integer.parseInt(String.valueOf(uncap.charAt(7)));
+                        response = tasks.unmark(i, ui);
+                    } else if (uncap.startsWith("mark")) {
```
Deeply nested if else loops are quite confusing, perhaps could make use of switch statements instead?

paullee18 commented on src/main/java/dukechatbot/utility/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+        if (String.valueOf(ln.charAt(4)).equals("X")) {
+            done = true;
+        }
+        if (tag.equals("T")) {
```
Perhaps this parsing of the command to determine how to create the Todo should be done in the Parser class?

paullee18 commented on src/main/java/dukechatbot/utility/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+        String tag = String.valueOf(ln.charAt(1));
+        String desc = null;
+        int id = -1;
+        boolean done = false;
```
Could use isDone as variable name instead

paullee18 reviewed (commented):
Overall good but some minor coding standard violations.

yuanxi1 commented on src/main/java/dukechatbot/duke/Duke.java:
```diff
@@ -0,0 +1,98 @@
+        } catch (IOException io) {
+            System.out.println("--------------------------------------\n");
+            System.out.println(io.getMessage());
+            System.out.println("--------------------------------------\n");
```
Maybe this can be abstracted into a utility method, for example `formatErrorMessage`?

yuanxi1 commented on src/main/java/dukechatbot/utility/Ui.java:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    /**
+     * lists out the tasks in the task list to the user.
```
This is not stated explicitly in the coding standard docs but I think the first letter should be capitalised?

yuanxi1 commented on src/main/java/dukechatbot/utility/Ui.java:
```diff
@@ -0,0 +1,162 @@
+    /**
+     * lists out the matching tasks in the task list that meets
+     * the user's find query.
+     * @param tl containing the tasks store in the task list.
```
Could add an empty line between description and parameter section.

yuanxi1 commented on src/main/java/dukechatbot/utility/Storage.java:
```diff
@@ -0,0 +1,83 @@
+            } else {
+                throw new IOException("File failed creation!");
+            }
+        }
```
Since if the named file already exits `createNewFile()` does nothing and return false, maybe this part can be further simplified? \
Besides, the `IOException` will be thrown if the file already exits, if that's the intended behaviour maybe it would be clearer if the error message indicate that file already exits?

yuanxi1 reviewed (commented):
Overall, I think your code is clean and easy to understand!

Lan-Jingbo reviewed (commented):
Very good

jwdavis0200 committed:
Change save method into static

save was a non-static method of the Storage class.

The save() should be static method since it's associated to Storage class not the instance.

Changed save to be static and modified code to work with this change in implementation.

jwdavis0200 committed:
Modify build.gradle

Build.gradle does not point to Launcher class which meant the jar file does not use javaFX.

To create a Jar file using JavaFX, we need to point to the Launcher class.

Modify build.gradle for javaFX implementation.

jwdavis0200 committed:
Add Assert statements.

The program has no way of testing for errors during the runtime.

Assert statements can help with this by forcing the program to terminate on bug.

Added Assert statements in some classes to check for runtime error that could have been caused by programmer error.

jwdavis0200 committed:
Merge pull request #3 from jwdavis0200/branch-Assertions

Add Assert statements.

jwdavis0200 committed:
Improve Code Quality

The Java classes have poor coding quality standards.

To improve readability, it is imperative for us to fix this.

Improved readability by fixing issues such as deep nested conditionals
and extracting methods into helper methods amongst other tweaks.

jwdavis0200 committed:
Improve Code Quality

The Java classes have poor coding quality standards.

To improve readability, it is imperative for us to fix this.

Improved readability by fixing issues such as deep nested conditionals
and extracting methods into helper methods amongst other tweaks.

jwdavis0200 committed:
Merge pull request #4 from jwdavis0200/branch-CodingStandard

Improve Code Quality

jwdavis0200 committed:
Add duplicate detection feature to Duke.

Duplicate tasks can be added into the Duke task list.

This may be a mistake by the user and may cause confusion.

Added the methods isDuplicate to check for duplicates, modified
the add method and added a method into UI class to implement
the feature to prevent duplicate tasks when user tries to add duplicates.

jwdavis0200 committed:
Update README.md

jwdavis0200 committed:
Update README.md

jwdavis0200 committed:
Modify README to suit product for user guide.

Old template of Duke is for original template from upstream repo.

There are alot of updates to Duke with added functionalities to support new features.

Updated README for users to learn how to use the new features implemented into Duke.

jwdavis0200 committed:
Merge branch 'master' of https://github.com/jwdavis0200/ip_own

jwdavis0200 committed:
Modify to fix file reading bug

BufferedReader was placed in the wrong line and may read file before file exists.

This can cause the program to fail when user runs an exported jar file.

Modifie lines of code to deal with this issue.

jwdavis0200 committed:
Update Readme

Readme was the plain template providede by the greenfield package.

We should update the Readme to accurately reflect the features of Duke.

Updated the Readme to reflect Duke features and how users cana use Duke.

jwdavis0200 committed:
Modify the guard clauses of Storage constructor.

The Storage constructor has extra try catch block for IOException handling.

This is not required with the modified logic of the constructor.

jwdavis0200 committed:
Remove HTML tracking.

Previously tracked all HTML updates.

HTML files are easily generated when needed for javadocs.

Removed HTML tracing via gitignore

jwdavis0200 committed:
Move README.md

jwdavis0200 committed:
Update build.gradle for name of archive

Name of jar will be duke.

Needs the jar build to be named DukeTaskBot.

Changed archive build name by shadowjar plugin to be DukeTaskBot
